client/src/model/values/instantaneousValues/vectorValues/VectorValue.hx

```diff
@@ -0,0 +1,148 @@
+package model.values.instantaneousValues.vectorValues;
+import model.values.instantaneousValues.vectorValues.VectorValueI;
+import model.values.instantaneousValues.InstantaneousValueI;
+import model.values.instantaneousValues.scalarValues.ScalarValueI;
+
+/**
+ * ...
+ * @author AdvaitTrivedi
+ */
+class VectorValue implements VectorValueI 
+{
+	var vector : Array<InstantaneousValueI> = new Array<InstantaneousValueI>();
+
+	public function new() 
+	{
+		
+	}
+	
+	
+	/* INTERFACE model.values.VectorValueI */
+	
+	@:allow(model.values.SignalValue)
```

Comment:
You could just make the constructor take an array as a parameter and then have no mutators in this class.

Reply:
But do we not want the ability to mutate signals, which will be used by the developer, or by the user (while making changes to the UI, say he extends the signal or cuts it short?) further more?

Reply:
Once a VectorValue is made it will be immutable.  E.g. you would never got something like this.

````
     VectiorValue vv = signalValue.getValueAtTime( t ) ;
     vv.push( ... ) 
     signalValue.setValueAtTimes( t0, t1, vv ) ;
````

because you have no idea where else that vector value object is being used.

What you might do is this

````
        VectiorValue vv = signalValue.getValueAtTime( t ) ;
       VectorValue newVV = vv.clone() ;
       newVV.push( ... )
      signalValue.setValueAtTimes( t0, t1, newVV ) ;
````

That would be ok since the clone method would make a completely new object that you know isn't being used anywhere else.  But an even better way might be 

````
       VectiorValue vv = signalValue.getValueAtTime( t ) ;
       MutableVectorValue mvv = new MutableVectorValue( vv ) ;
       newVV.push( ... )
      signalValue.setValueAtTimes( t0, t1, mvv.get() ) ;
````

Here MutableVectorValue and VectorValue have roughly the same relationship as StringBuilder and String in Java or Set and Frozenset in Python.  I.e., the second is immutable and the first is a builder for the second that can be initialized by the first.  It's a form of the Builder pattern.  This approach let static typechecking verify the integrity.

A middle ground is to give the VectorValue a "frozen" attribute. When frozen is true, the vector's modifiers may not be called (which you can check at run time with `Assert.assert( ! this.isFrozen() )`.  Once frozen, a vector can not be unfrozen.  Then it is a precondition to methods that expect frozen values that the value is frozen. setValueAtTimes would be such a method.  It's first line would be `Assert.assert( vv.isFrozen() )`.  In this case client code might look like this.

````
       VectiorValue vv = signalValue.getValueAtTime( t ) ;
       MutableVectorValue mvv = new MutableVectorValue( vv ) ;
       newVV.push( ... ) ;
       newVV.freeze() ;
      signalValue.setValueAtTimes( t0, t1, mvv.get() ) ;
````

Reply:
If we go with this path then the Vector is better off as immutable.
I shall now remove all mutability in a vector.